javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1425,4 +1359,242 @@ object Formatter {
+        // FINITE STATE MACHINE
+        val prevChar = currentChar
+        if (ParserStateMachine.Exit != state) {
+          // exit state does not need to get next char
+          currentChar = getNextFormatChar()
+          if (EOS == currentChar && ParserStateMachine.Entry != state) {
+            throw new UnknownFormatConversionException("%")
+          }
+        }
+
+        (state: @switch) match {
+          case ParserStateMachine.Exit =>
+            token.plainText = getFormatString(token)
+            token
+
+          case ParserStateMachine.Entry =>
+            loop {
+              if (EOS == currentChar) ParserStateMachine.Exit
+              else if ('%' == currentChar) {
+                token.formatSpecifierIndex =
+                  format.position() - token.formatStringStartIndex
+                StartConversion
+              } else state
+            }
+
+          case ParserStateMachine.StartConversion =>
+            loop {
+              if (Character.isDigit(currentChar)) {
+                val position       = format.position() - 1
+                val number         = getPositiveInt(format)
+                var nextChar: Char = 0
+                if (format.hasRemaining()) {
+                  nextChar = format.get()
+                }
+
+                val newState = if ('$' == nextChar) {
+                  token.argIndex = number
+                  ApplyFlags
+                } else {
+                  // the digital zero stands for one format flag.
+                  if ('0' == currentChar) {
+                    format.position(position)
+                    ApplyFlags
+                  } else {
+                    // the digital sequence stands for the width.
+                    // do not get the next char.
+                    format.position(format.position() - 1)
+                    token.width = number
+                    ApplyWidth
+                  }
+                }
+                currentChar = nextChar
+                newState
+              } else {
+                if ('<' == currentChar) {
+                  token.argIndex = FormatToken.LastArgumentIndex
+                } else {
+                  // do not get the next char.
+                  format.position(format.position() - 1)
+                }
+                ApplyFlags
+              }
+            }
+
+          case ParserStateMachine.ApplyFlags =>
+            loop {
+              if (token.setFlag(currentChar)) state
+              else if (Character.isDigit(currentChar)) {
+                token.width = getPositiveInt(format)
+                ApplyWidth
+              } else if ('.' == currentChar) {
+                ApplyPrecision
+              } else {
+                // do not get the next char.
+                format.position(format.position() - 1)
+                ApplyConversionType
+              }
+            }
+
+          case ParserStateMachine.ApplyWidth =>
```

Comment:
It looks like this state should be called `AfterWidth` rather than `ApplyWidth`. The `Apply*` state are the *start* of the respective `*`s.